contracts/AddressBook.sol

```diff
@@ -3,10 +3,233 @@
+        return getAddress(WETH);
+    }
+
+    /**
+     * @notice set a specific address for a specific key
+     * @dev can only be called by addressbook owner
+     * @param _key key
+     * @param _address address
+     */
+    function setAddress(bytes32 _key, address _address) public onlyOwner {
+        addresses[_key] = _address;
+
+        emit AddressAdded(_key, _address);
+    }
+
+    /**
+     * @notice set otoken implementation address
+     * @dev can only be called by addressbook owner
+     * @param _otokenImpl otoken implementation address
+     */
+    function setOtokenImpl(address _otokenImpl) external onlyOwner {
+        setAddress(OTOKEN_IMPL, _otokenImpl);
+    }
+
+    /**
+     * @notice set otoken factory address
+     * @dev can only be called by addressbook owner
+     * @param _otokenFactory otoken factory address
+     */
+    function setOtokenFactory(address _otokenFactory) external onlyOwner {
+        setAddress(OTOKEN_FACTORY, _otokenFactory);
+    }
+
+    /**
+     * @notice set whitelist address
+     * @dev can only be called by addressbook owner
+     * @param _whitelist whitelist address
+     */
+    function setWhitelist(address _whitelist) external onlyOwner {
+        setAddress(WHITELIST, _whitelist);
+    }
+
+    /**
+     * @notice set controller address
+     * @dev can only be called by addressbook owner
+     * @param _controller controller address
+     */
+    function setController(address _controller) external onlyOwner {
+        updateImpl(CONTROLLER, _controller);
+    }
+
+    /**
+     * @notice set pool address
+     * @dev can only be called by addressbook owner
+     * @param _marginPool pool address
+     */
+    function setMarginPool(address _marginPool) external onlyOwner {
+        setAddress(MARGIN_POOL, _marginPool);
+    }
+
+    /**
+     * @notice set margin calculator address
+     * @dev can only be called by addressbook owner
+     * @param _marginCalculator margin calculator address
+     */
+    function setMarginCalculator(address _marginCalculator) external onlyOwner {
+        updateImpl(MARGIN_CALCULATOR, _marginCalculator);
+    }
+
+    /**
+     * @notice set liquidation manager address
+     * @dev can only be called by addressbook owner
+     * @param _liquidationManager liquidation manager address
+     */
+    function setLiquidationManager(address _liquidationManager) external onlyOwner {
+        updateImpl(LIQUIDATION_MANAGER, _liquidationManager);
+    }
+
+    /**
+     * @notice set oracle address
+     * @dev can only be called by addressbook owner
+     * @param _oracle oracle address
+     */
+    function setOracle(address _oracle) external onlyOwner {
+        updateImpl(ORACLE, _oracle);
+    }
+
+    /**
+     * @notice set WETH address
+     * @dev can only be called by addressbook owner
+     * @param _weth weth address
+     */
+    function setWeth(address _weth) external onlyOwner {
+        setAddress(WETH, _weth);
+    }
```

Comment:
should this be here or in the whitelist module?

Reply:
It should be here as the `AddressBook` is the addresses provider for the system.